src/server/pps/cmds/cmds.go

```diff
@@ -1442,53 +1442,72 @@ All jobs created by a pipeline will create commits in the pipeline's output repo
 	nextDatum.Flags().StringVar(&errStr, "error", "", "A string representation of an error that occurred while processing the current datum.")
 	commands = append(commands, cmdutil.CreateAlias(nextDatum, "next datum"))
 
+	validatePipeline := &cobra.Command{
+		Use:   "{{alias}}",
+		Short: "Validate pipeline spec.",
+		Long:  "Validate a pipeline spec.  Client-side only; does not check that repos, images &c. exist on the server.",
+		Run: cmdutil.RunFixedArgs(0, func(_ []string) error {
+			r, err := fileIndicatorToReader(pipelinePath)
+			if err != nil {
+				return err
+			}
+			pr, err := ppsutil.NewPipelineManifestReader(r)
+			if err != nil {
+				return err
+			}
+			for {
+				_, err := pr.NextCreatePipelineRequest()
+				if errors.Is(err, io.EOF) {
+					return nil
+				} else if err != nil {
+					return err
+				}
+			}
+		}),
+	}
+	validatePipeline.Flags().StringVarP(&pipelinePath, "file", "f", "", "A JSON file (url or filepath) containing one or more pipelines. \"-\" reads from stdin (the default behavior). Exactly one of --file and --jsonnet must be set.")
+	commands = append(commands, cmdutil.CreateAliases(validatePipeline, "validate pipeline", pipelines))
+
 	return commands
 }
 
-// readPipelineBytes reads the pipeline spec at 'pipelinePath' (which may be
-// '-' for stdin, a local path, or a remote URL) and returns the bytes stored
-// there.
+// fileIndicatorToReader returns an IO reader for a file based on an indicator
+// (which may be a local path, a remote URL or "-" for stdin).
 //
 // TODO(msteffen) This is very similar to readConfigBytes in
 // s/s/identity/cmds/cmds.go (which differs only in not supporting URLs),
 // so the two could perhaps be refactored.
-func readPipelineBytes(pipelinePath string) (pipelineBytes []byte, retErr error) {
-	if pipelinePath == "-" {
+func fileIndicatorToReader(indicator string) (io.Reader, error) {
```

Comment:
could this be called `filePathToReader`?

Reply:
I think that would be incorrect, because it accepts more than just paths: literal `"-"` is interpreted as STDIN; a string which parses as a URL is interpreted as a URL; only all other strings are interpreted as paths.

Reply:
Not a big deal but "-" is just a conventional "path" standin for terminal I/O offered by commandline tools - does it really need to be supported in a path parameter in an internal API?

Could something like "file:///dev/stdin" or "file:///dev/fd/0" be more clear and consistent?

Reply:
It’s not supported in the internal API, just in the `pachctl` command-line tool. Our code organization is misleading: it looks like this is server code (because it’s under `src/server`), but really it is client code (because it’s under `src/server/pps/cmd`: yes, the client code for each subsystem is located beneath the server directory). IMHO, at some point we should extract all the stuff in `src/server/.../cmds/` into something under `src/cmd/...`, but that’s another story☺